src/dialects/abstract/data-types.ts

```diff
@@ -0,0 +1,1826 @@
+        ),
+        [],
+      );
+    }
+
+    return true;
+  }
+
+  protected _stringify(number: AcceptedNumber): string {
+    // This should be unnecessary but since this directly returns the passed string its worth the added validation.
+    this.validate(number);
+
+    return String(number);
+  }
+
+  public get UNSIGNED() {
+    return this._construct<typeof NUMBER>({ ...this.options, unsigned: true });
+  }
+
+  public get ZEROFILL() {
+    return this._construct<typeof NUMBER>({ ...this.options, zerofill: true });
+  }
+
+  public static get UNSIGNED() {
+    return new this({ unsigned: true });
+  }
+
+  public static get ZEROFILL() {
+    return new this({ zerofill: true });
+  }
+}
+
+/**
+ * Base number type which is used to build other types
+ */
+export type NUMBER = _NUMBER;
+export const NUMBER = classToInvokable(_NUMBER);
+
+class _INTEGER extends NUMBER {
+  public static readonly key: string = 'INTEGER';
+
+  public validate(value: AcceptedNumber) {
+    if (!Validator.isInt(String(value))) {
+      throw new ValidationError(
+        util.format(`%j is not a valid ${this.key.toLowerCase()}`, value),
+        [],
+      );
+    }
+
+    return true;
+  }
+}
+
+/**
+ * A 32 bit integer
+ */
+export type INTEGER = _INTEGER;
+export const INTEGER = classToInvokable(_INTEGER);
+
+class _TINYINT extends INTEGER {
+  public static readonly key = 'TINYINT';
+}
+
+/**
+ * A 8 bit integer
+ */
+export type TINYINT = _TINYINT;
+export const TINYINT = classToInvokable(_TINYINT);
+
+class _SMALLINT extends INTEGER {
+  public static readonly key = 'SMALLINT';
+}
+
+/**
+ * A 16 bit integer
+ */
+export type SMALLINT = _SMALLINT;
+export const SMALLINT = classToInvokable(_SMALLINT);
+
+class _MEDIUMINT extends INTEGER {
+  public static readonly key = 'MEDIUMINT';
+}
+
+/**
+ * A 24 bit integer
+ */
+export type MEDIUMINT = _MEDIUMINT;
+export const MEDIUMINT = _MEDIUMINT;
+
+class _BIGINT extends INTEGER {
+  public static readonly key = 'BIGINT';
+}
+
+/**
+ * A 64 bit integer
+ */
+export type BIGINT = _BIGINT;
+export const BIGINT = classToInvokable(_BIGINT);
+
+type FloatConstructorParams =
+  | []
+  | [NumberOptions | undefined]
+  | [number | undefined]
+  | [number | undefined, number | undefined];
+
+class _FLOAT extends NUMBER {
+  public static readonly key: string = 'FLOAT';
+  public static readonly escape = false;
+
+  /**
+   * @param {object|string|number} [length] length of type, like `FLOAT(4)`
+   * @param {string|number} [decimals] number of decimal points, used with length `FLOAT(5, 4)`
+   */
+  constructor(options: NumberOptions);
+  constructor(length: number, decimals?: number);
+  constructor(...args: FloatConstructorParams);
+  constructor(length?: number | NumberOptions, decimals?: number) {
+    super(typeof length === 'object' ? length : { length, decimals });
+  }
+
+  public validate(value: AcceptedNumber): asserts value is AcceptedNumber;
+  public validate(value: AcceptedNumber): true {
```

Comment:
can't this be simplified to:

```suggestion
  public validate(value: AcceptedNumber): asserts value is AcceptedNumber {
```